Updates to `@_predatesConcurrency`

DougGregor opened the pull request:
Implement a few remaining pieces of the `@_predatesConcurrency` attribute:

* Adjust the mangled type of `@_predatesConcurrency` declarations to drop `Sendable`, `@Sendable`, and global actor annotations
* All imported C declarations are `@_predatesConcurrency` by default

DougGregor committed:
Imported C declarations always predate concurrency

DougGregor committed:
Strip concurrency-related types from mangled name of @_predatesConcurrency decls

This allows some APIs to evolve toward supporting concurrency without
breaking their ABI.

DougGregor commented:
@swift-ci please build toolchain

beccadax commented on lib/AST/Decl.cpp:
```diff
@@ -719,7 +720,14 @@ Optional<CustomAttrNominalPair> Decl::getGlobalActorAttr() const {
+  if (getAttrs().hasAttribute<PredatesConcurrencyAttr>())
+    return true;
+
+  // Imported C declarations always predate concurrency.
+  if (isa<ClangModuleUnit>(getDeclContext()->getModuleScopeContext()))
+    return true;
```
Out of curiosity, what will this match that `hasClangNode()` wouldn't or vice versa?

DougGregor replied:
IIRC, there's some cases involving synthesized extensions for import-as-member that differ.

beccadax commented on lib/AST/Type.cpp:
```diff
@@ -755,6 +755,94 @@ Type TypeBase::lookThroughAllOptionalTypes(SmallVectorImpl<Type> &optionals){
+    // Strip off Sendable and (possibly) the global actor.
+    ASTExtInfo extInfo =
+        fnType->hasExtInfo() ? fnType->getExtInfo() : ASTExtInfo();
+    extInfo = extInfo.withConcurrent(false);
```
`withConcurrent()` should probably get renamed eventually, but not today.

beccadax commented on include/swift/AST/ASTMangler.h:
```diff
@@ -67,6 +67,9 @@ class ASTMangler : public Mangler {
   /// If enabled, marker protocols can be encoded in the mangled name.
   bool AllowMarkerProtocols = true;
 
+  /// Whether the mangling predates concurrency, and therefore shouldn't
```
I think you didn't finish typing this.

beccadax reviewed (approved):
A nit and a question, but this looks good.

DougGregor committed:
Eliminate support for @MainActor(unsafe) Clang attribute.

The `@MainActor(unsafe)` attribute could be provided for C declarations
via the Clang `swift_attr` attribute. However, this facility was never
used outside of tests, and has been superceded by `@MainActor` with the
inferred `@_predatesConcurrency`.

DougGregor committed:
Finish a comment

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test
